src/Microsoft.AspNetCore.OData/EnableQueryAttribute.cs

```diff
@@ -38,6 +41,170 @@ public partial class EnableQueryAttribute : ActionFilterAttribute
         // also clearly ties the property to code in this assembly.
         private const string ModelKeyPrefix = "Microsoft.AspNet.OData.Model+";
 
+        /// <summary>
+        /// Marks if the query validation was run before the action execution. This is not always possible.
+        /// For cases where the run failed before action execution. We will run validation on result.
+        /// </summary>
+        private bool _queryValidationRunBeforeActionExecution;
+
+        /// <summary>
+        /// Stores the processed query options to be used later if OnActionExecuting was able to verify the query.
+        /// This is because ValidateQuery internally modifies query options (expands are prime example of this).
+        /// </summary>
+        private ODataQueryOptions _processedQueryOptions;
+
+        /// <summary>
+        /// Performs query validations before action is executed.
+        /// </summary>
+        /// <param name="context">Action context.</param>
+        public override void OnActionExecuting(ActionExecutingContext context)
+        {
+            if (context == null)
+            {
+                throw Error.ArgumentNull(nameof(context));
+            }
+
+            base.OnActionExecuting(context);
+
+            HttpRequest request = context.HttpContext.Request;
+            ODataPath path = request.ODataFeature().Path;
+
+            ODataQueryContext queryContext = null;
+
+            // For OData based controllers.
+            if (path != null)
+            {
+                IEdmType edmType = path.EdmType;
+
+                // When $count is at the end, the return type is always int. Trying to instead fetch the return type of the actual type being counted on.
+                if (path.Segments.Last().Identifier == "$count")
+                {
+                    edmType = path.Segments[path.Segments.Count - 2].EdmType;
+                }
```

Comment:
It seems we have the similar extension method test whether it's $count request?

Reply:
Can you point me to it?

Reply:
public static bool IsCountRequest(this HttpRequest request)

---
In reply to: [538873066](https://github.com/OData/WebApi/pull/2362#discussion_r538873066) [](ancestors = 538873066)

Reply:
Done.